src/tools/perf/lib/ucp_tests.cc

```diff
@@ -120,6 +124,39 @@ class ucp_perf_test_runner {
         }
     }
 
+    void ucp_perf_init_common_params(size_t *total_length, size_t *send_length,
+                                     ucp_datatype_t *send_dt,
+                                     void **send_buffer, size_t *recv_length,
+                                     ucp_datatype_t *recv_dt,
+                                     void **recv_buffer)
+    {
+        *total_length = ucx_perf_get_message_size(&m_perf.params);
+
+        ucs_assert(*total_length >= sizeof(psn_t));
+
+        ucp_perf_test_prepare_iov_buffers();
+
+        *send_length = *recv_length = *total_length;
+
+        *send_dt = ucp_perf_test_get_datatype(m_perf.params.ucp.send_datatype,
+                                              m_perf.ucp.send_iov, send_length,
+                                              send_buffer);
+        *recv_dt = ucp_perf_test_get_datatype(m_perf.params.ucp.recv_datatype,
+                                              m_perf.ucp.recv_iov, recv_length,
+                                              recv_buffer);
+        if (CMD == UCX_PERF_CMD_AM) {
+            m_am_rx_params.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
+                                          UCP_OP_ATTR_FIELD_USER_DATA |
+                                          UCP_OP_ATTR_FIELD_DATATYPE |
+                                          UCP_OP_ATTR_FLAG_NO_IMM_CMPL;
+            m_am_rx_params.datatype     = *recv_dt;
+            m_am_rx_params.cb.recv_am   = am_data_recv_cb;
+            m_am_rx_params.user_data    = this;
+            m_am_rx_buffer              = *recv_buffer;
+            m_am_rx_length              = *recv_length;
```

Comment:
why need m_am_rx_buffer,m_am_rx_length - can't just use m_perf.recv_buffer/m_perf.recv_length?

Reply:
no, it can change in case of iov

Reply:
why is it relevant just for am, and not for tag for example?

Reply:
because in tag we pass recv buffer and length to the `recv` func in the benchmark. AM does not have direct `recv` call.
Since now we are precalculating datatypes, length, etc, would need to store rx_buf and length somewhere